test: sbs_gauge_new_api replace redundant overlays

aaronemassey opened the pull request:
Note: Includes a fix to the SBS compatible IC emulator.

The fuel gauge API sbs_gauge test uses conf/overlay pairs for each board. This was getting unnecessary and redundant.

Combine the overlay/conf files into a single generic overlay/conf pair for emulated boards and apply them via the testcase.yaml rules.

Motivated by https://github.com/zephyrproject-rtos/zephyr/pull/53632#pullrequestreview-1308170330

aaronemassey committed:
test: sbs_gauge_new_api replace redundant overlays

The fuel gauge API sbs_gauge test uses conf/overlay pairs for each
board. This was getting unnecessary and redundant.

Combine the overlay/conf files into a single generic overlay/conf pair for
emulated boards and apply them via the testcase.yaml rules.

Signed-off-by: Aaron Massey <aaronmassey@google.com>

teburd commented on tests/drivers/fuel_gauge/sbs_gauge/testcase.yaml:
```diff
@@ -3,6 +3,13 @@ tests:
+    extra_args:
+      CONF_FILE="prj.conf;boards/emulated_board.conf"
+      DTC_OVERLAY_FILE="boards/emulated_board.overlay"
+    platform_allow:
```
Droop the platform allow now that you have a common emulated overlay, should work everywhere.

aaronemassey replied:
```
INFO    - The following issues were found (showing the top 10 items):
INFO    - 1) tests/drivers/fuel_gauge/sbs_gauge/drivers.sbs_gauge_new_api on qemu_cortex_a53_smp error (Cmake build failure)
INFO    - 2) tests/drivers/fuel_gauge/sbs_gauge/drivers.sbs_gauge_new_api on qemu_cortex_a53 error (Cmake build failure)
INFO    - 3) tests/drivers/fuel_gauge/sbs_gauge/drivers.sbs_gauge_new_api on qemu_malta_be failed (Failed)
INFO    - 4) tests/drivers/fuel_gauge/sbs_gauge/drivers.sbs_gauge_new_api on qemu_leon3 failed (Failed)
INFO    - 
INFO    - To rerun the tests, call twister using the following commandline:
```

Running `twister -c -v -i -T tests/drivers/fuel_gauge/` seems to work on all but 3 (technically 4) qemu boards. That might take quite a bit of debugging which I'd prefer to do in some other PR with added TODO for these specifically. It looks like `qemu_cortex_a53*` boards seem to just need a special case overlay to build.

Do we have a way to disallow just these platforms? I could keep the `platform_allow` for all the ones that passed but that seems verbose.

teburd replied:
platform_exclude can exclude specific platforms, but personally I'd spend some time looking at the issues, they can be useful in catching problems with your API

aaronemassey replied:
It did catch a bug in my emulator :stuck_out_tongue:

aaronemassey committed:
emul: Fix endianness fault in sbs gauge emulator

The sbs gauge emulator intercepts SMBus messages for an sbs gauge
compatible IC. However, it was incorrectly interpreting the byte order of
received register write values based on the endianness of the SoC
architecture instead of SMBus' defined little-endianness. This fault was
not caught earlier because native posix and the majority of qemu
architectures use little-endian.

Explicitly interpret the write values as little-endian.

Signed-off-by: Aaron Massey <aaronmassey@google.com>

aaronemassey commented:
**tldr**: This PR is ready for review again after a minor change of special-casing qemu_cortex_a53 to maintain proper `reg` properties in overlays.

Okay in order to get rid of the devicetree build warning, namely `unit address and first address in 'reg' (0x0) don't match for /i2c@100` while getting `qemu_cortex_a53` to build I had to special case `qemu_cortex_a53` builds as its `/soc` node requires the `reg` property of its child i2c bus node to be made up of a 64 bit address of `<addr-high addr-low size-high size-low>`

Usually this wouldn't matter because neither the driver, tests, nor emulator make use of the `reg` property of the `i2c` node. However, @keith-zephyr pointed out the malformed `reg` property could be confusing to others using the overlay as an example and I agreed.

I'll be creating three GitHub issues:
* Devicetree build warnings should be surfaced by twister
* Qemu_x86_64 and other 64 bit qemus *maybe* should be using 64 bit reg addressing; this would've surfaced this problem earlier.
* FeatureRequest: Selecting boards based on properties like `#address-cells` for `/soc`.

teburd commented:
Seems like a similar 64bit reg property issue is still failing in CI for kvm_arm64

aaronemassey commented:
> Seems like a similar 64bit reg property issue is still failing in CI for kvm_arm64

Yeah I'm surprised I'm not running all the relevant platforms with a `./twister -T tests/fuel_gauge` Unfortunately it just means my special case has to include four more boards.

aaronemassey commented:
Turns out I was building with just the sdk and not the docker container which actually includes all of the qemus. Thanks @yperess !

teburd commented:
CI adds --all, so something like...

```
scripts/twister -c -T tests/drivers/fuel_gauge/sbs_gauge --all
```

Get your fan ready.
